### ISSUE TITLE
Support Interpol ICSE baseline CSAM hash database

### ISSUE THREAD

**Issue by lfcnassif** [sepinf-inc/IPED#1136]
Seems their baseline database has grown a lot, to ~2.7M files when this article was published:
https://www.interpol.int/Crimes/Crimes-against-children/International-Child-Sexual-Exploitation-database

I'm checking with @herrmannpchw how we can get access to an updated version and format documentation.

**Comment by lfcnassif** [comment]
Seems the format is CSV. I'm checking with Marcelo Abdalla for an updated version to test if our CSV importing works out of the box (thanks @tc-wleite!)

**Comment by lfcnassif** [comment]
I just downloaded ICSE hashset. Taking a quick look, it contains ~3.1M hashes, CSV format using semicolon, headers are:

MD5;SHA1;ZZ40;PhotoDNA;Lace;Image ID;Is distributed;Victim identified;Offender identified;Series name;Investigation name;Agency;Number of submissions;File Available;Media Type;Baseline

DataTypes:

MD5: lowercase hex
SHA1: lowercase hex
ZZ40: uppercase hex (128 bits)
PhotoDNA: uppercase hex (144 bytes)
Lace: base64 (129 bytes)
Image ID: string
Is distributed: Y|N
Victim identified: Y|N
Offender identified: Y|N
Series name: string
Investigation name: string
Agency: string
Number of submissions: integer
File Available: True|False
Media Type: Image|Video
Baseline: Yes|No

**Comment by lfcnassif** [comment]
So, possibly we will need to update a bit the code or convert the CSV file format, because we handle base64 encoded PhotoDNA today.

**Comment by lfcnassif** [comment]
Found some info about ZZ40 in internet history: https://web.archive.org/web/20210508171625/https://www.ziuz.com/industries/law-enforcement/technology/

**Comment by lfcnassif** [comment]
And Lace is from BlueBear: https://bb-les.ca/solutions/

**Comment by lfcnassif** [comment]
> Investigation name: string
> Agency: string

These are sensitive information and Marcelo Abdalla recommended removing those fields before importing this hashset into our internal hash database, since that would be included in reports (accessed by third parties). Any objections?

**Comment by lfcnassif** [comment]
> It is possible to write a small conversion program that would take care of these columns, change the column separator, add "set" and "status" columns, and convert PhotoDNA format?

I think so. @tc-wleite what do you think about `ZZ40 `and `Lace` hashes, since we don't support them today (and I don't know if they will be supported in the future since I _think_ they are commercial), should they be also removed?

**Comment by tc-wleite** [comment]
Another solution would be changing importing tool to allow other command line parameters, like separator char, options to ignore columns (could be used for the ones you mentioned) and to add fixed columns (like "set"). 
For the "status" it would be so simple, as there could be different ones in the dataset. But there could be a mapping option... 
PhotoDNA is also simple to handle both hex and base64 encodings. 

I thinking about a more general solution, at least to allow some flexibility.

**Comment by tc-wleite** [comment]
@lfcnassif, what do you think is the best option among the discussed ones:

1. Change import tool to handle this file format.
2. Write (and share) a simple converter.
3. Change import tool to handle more situations (separators, skip/add/rename/map columns).

The first option would make the import process easier, but it would be a solution for this particular format only.

In the third option, a lot of possible parameters could be implemented, but my idea would be adding only the basic ones (basically, enough to handle this file). 
One disadvantage would be that the user would need to use a certain combination of parameters to read the ICSE file.
The exact command could be added to the project's Wiki. 

By the way, can you send me this file?
I would like to check the intersection with ProjectVIC and other hashsets we use.

**Comment by lfcnassif** [comment]
Hi @tc-wleite I think option 3 would be a more general solution, but I'm ok with any of them. But option 1 would avoid our colleagues from importing `agency` and `investigation` columns by accident...

> By the way, can you send me this file?

For sure! I'll send you privatelly by Teams.

**Comment by lfcnassif** [comment]
Thinking better, option 1 would avoid importing and distributing sensitive information, by internal and external users.

**Comment by tc-wleite** [comment]
Good point!
What about implementing the first option now and left the third as an independent (future?) improvement for other CSV files?

**Comment by lfcnassif** [comment]
That would be great!

**Comment by tc-wleite** [comment]
Starting implementing the **second option**, so we can have a version of the file without some columns and with the default separator (comma). 
The following column names mapping was applied, trying to have column names (prefix/case) similar to the ones used by ProjectVIC, NSRL and other hash sets already supported.
Boolean values were also normalized to from ("Y", "N", "True", "False", "T", "F", "Yes", "No") to ("true", "false").

|Original|Output CSV|
|:---|:---|
|MD5|MD5|
|SHA1|SHA1|
|ZZ40|-|	
|PhotoDNA|photoDNA|
|Lace|-|
|Image ID|icseImageID|
|Is distributed|icseIsDistributed|
|Victim identified|icseVictimIdentified|
|Offender identified|icseOffenderIdentified|
|Series name|icseSeriesName|
|Investigation name|-|	
|Agency|-|	
|Number of submissions|icseNumSubmissions|
|File Available|icseFileAvailable|
|Media Type|icseMediaType|
|Baseline|icseBaseline|
|-|status (fixed as "pedo")|
|-|set (fixed as "ICSE")|

**Comment by lfcnassif** [comment]
Excellent, thank you @tc-wleite for that comparison! I think it will improve our internal hash database a lot, although any large hashset is vulnerable to false positives...

**Comment by lfcnassif** [comment]
I _think_ `icseBaseline ` would be something similar to VIC category 1

**Comment by tc-wleite** [comment]
A simple code to convert from the orignal format to a standard format already supported by IPED's HashDB importing tool.

```java
import java.io.*;
import java.util.*;

public class IcseToCsv {
    public static void main(String[] args) throws Exception {
        File inputFile = new File(args[0]);
        File outputFile = new File(args[1]);
        BufferedReader in = new BufferedReader(new FileReader(inputFile), 1 << 20);
        BufferedWriter out = new BufferedWriter(new FileWriter(outputFile), 1 << 20);
        StringBuilder sb = new StringBuilder();
        sb.append("MD5,");
        sb.append("SHA1,");
        //sb.append("ZZ40,");
        sb.append("photoDNA,");
        //sb.append("LACE,");
        sb.append("icseImageID,");
        sb.append("icseIsDistributed,");
        sb.append("icseVictimIdentified,");
        sb.append("icseOffenderIdentified,");
        sb.append("icseSeriesName,");
        //sb.append("InvestigationName,");
        //sb.append("Agency,");
        sb.append("icseNumSubmissions,");
        sb.append("icseFileAvailable,");
        sb.append("icseMediaType,");
        sb.append("icseBaseline,");
        sb.append("status,");
        sb.append("set");
        out.write(sb.toString());
        out.newLine();
        String line = in.readLine();
        while ((line = in.readLine()) != null) {
            String[] s = line.split(";");
            String md5 = s[0].trim();
            if (md5.length() != 0 && md5.length() != 32) continue;
            String sha1 = s[1].trim();
            if (sha1.length() != 0 && sha1.length() != 40) continue;
            if (md5.isEmpty() && sha1.isEmpty()) continue;
            sb.delete(0, sb.length());
            sb.append(md5).append(','); //MD5
            sb.append(sha1).append(','); //SHA1
            //sb.append(s[2]).append(','); //ZZ40
            sb.append(hexToBase64(s[3])).append(','); //PhotoDNA
            //sb.append(s[4]).append(','); //LACE
            sb.append(s[5]).append(','); //ImageID
            sb.append(norm(s[6])).append(','); //IsDistributed
            sb.append(norm(s[7])).append(','); //VictimIdentified
            sb.append(norm(s[8])).append(','); //OffenderIdentified
            sb.append(escape(s[9])).append(','); //SeriesName
            //sb.append(escape(s[10])).append(','); //InvestigationName
            //sb.append(escape(s[11])).append(','); //Agency
            sb.append(s[12]).append(','); //NumSubmissions
            sb.append(norm(s[13])).append(','); //FileAvailable
            sb.append(s[14].toLowerCase()).append(','); //MediaType
            sb.append(norm(s[15])).append(','); //Baseline
            sb.append("pedo,"); //status
            sb.append("ICSE"); //set
            out.write(sb.toString());
            out.newLine();
        }
        in.close();
        out.close();
    }

    private static String escape(String s) {
        s = s.replace('"', '\'');
        if (s.indexOf(',') >= 0) s = '"' + s + '"';
        return s;
    }

    private static String norm(String s) {
        if (s.equalsIgnoreCase("true")) return "true";
        if (s.equalsIgnoreCase("false")) return "false";
        if (s.equalsIgnoreCase("t")) return "true";
        if (s.equalsIgnoreCase("f")) return "false";
        if (s.equalsIgnoreCase("yes")) return "true";
        if (s.equalsIgnoreCase("no")) return "false";
        if (s.equalsIgnoreCase("y")) return "true";
        if (s.equalsIgnoreCase("n")) return "false";
        return s;
    }

    private static String hexToBase64(String s) {
        byte[] bytes = new byte[s.length() / 2];
        for (int i = 0; i < bytes.length; i++) {
            bytes[i] = (byte) Integer.parseInt(s.substring(i * 2, i * 2 + 2), 16);
        }
        return Base64.getEncoder().encodeToString(bytes);
    }
}
```

**Comment by tc-wleite** [comment]
> I _think_ `icseBaseline ` would be something similar to VIC category 1 (just a comment, I don't think it should be handled in a special way)

I also guessed that and tried to figure out if it could be used to avoid false positives.
But checking the file content, there were only ~40K with 'baseline' = 'Yes', so this column will be imported, but I think it can't be used in a straightforward way, as most hashes are not marked as baseline.

**Comment by tc-wleite** [comment]
> (...) although any large hashset is vulnerable to false positives...

That is true! In every release of LED I remove many false positives.

**Comment by tc-wleite** [comment]
> I will try to run later against a few real cases and see what kind and how many "exclusive hits" (not in VIC, LED) are found.

@lfcnassif, a quick update on this. 
Processing evidences from 4 different cases, many exclusive hits were found, but there were too many hits that are clearly "young adults". Out of ~20K hits images and videos only present in ICSE hash set (not in VIC nor LED), I estimated that at least 1/4 (~5K) were not CSAM. 

I still found useful to use this data set, but I am wondering if we should set `status: pedo` for it.
Not setting the status would exclude hits from pre-built filters and other useful functions, which doesn't seem a good idea.
It would still be possible to filter hits using `set: ICSE`, avoiding false positives, but less convenient. 
Well, we can always change this later, but what do you @lfcnassif?

**Comment by lfcnassif** [comment]
Hum that's not good... Maybe a different `status: pedo-suspect` or even `status:csam-suspect` (replacing the not so good `pedo` term in other places...)?
And about ProjectVIC category 2, it is being mapped to `status:pedo` right? Maybe that could be renamed too...

**Comment by tc-wleite** [comment]
> Hum that's not good... Maybe a different `status: pedo-suspect` or even `status:csam-suspect` (replacing the not so good `pedo` term in other places...)? 

I don't like the term 'pedo' neither. Replacing it seems a good idea, although many users are used to it.
Not sure if '-' could be used in the status name. Otherwise filtering by `status: csam` would bring `status: csam-suspect` too, right?

> And about ProjectVIC category 2, it is being mapped to `status:pedo` right? Maybe that could be renamed too...

Yes, hashes from VIC category 2 also receive 'status: pedo'. The problem here is that in practice, category 1 and 2 separation is not so clear. There are many images classified as 1 that could be 2 and the opposite. So I am not sure if another status would be helpful. It is already possible to use an explicit filter like `vicCategory: 1`.

**Comment by tc-wleite** [comment]
By the way, `csam` could be a replacement for `pedo` status. 
Although its usage is not common (at all) here in Brazil, I noticed that is commonly used in other places.

**Comment by lfcnassif** [comment]
> Not sure if '-' could be used in the status name. Otherwise filtering by `status: csam` would bring `status: csam-suspect` too, right?

Good point. We could remove the hyphen or, if some separator is desired, maybe fix the property to be indexed as StringField, not as Text, so it won't be tokenized.

**Comment by tc-wleite** [comment]
> Good point. We could remove the hyphen or, if some separator is desired, maybe fix the property to be indexed as StringField, not as Text, so it won't be tokenized.

Well, that was just a minor observation.
In fact, I am not sure if we should add more granularity to the `status` property.

**Comment by tc-wleite** [comment]
> Processing evidences from 4 different cases, many exclusive hits were found, but there were too many hits that are clearly "young adults". Out of ~20K hits images and videos only present in ICSE hash set (not in VIC nor LED), I estimated that at least 1/4 (~5K) were not CSAM.
 
Just to be clear, I am not saying that we should expect a similar rate of false positives in other cases.
Although I used files from 4 unrelated cases, it is a small sample.

**Comment by lfcnassif** [comment]
> The problem here is that in practice, category 1 and 2 separation is not so clear. There are many images classified as 1 that could be 2 and the opposite. So I am not sure if another status would be helpful. It is already possible to use an explicit filter like `vicCategory: 1`.

Agreed. Is there a significant number of adult porn in VIC category 2? Or most of them are really sequences or age difficult? If clear porn adult is already founf in VIC 2, we already have that ICSE issue happening...

**Comment by tc-wleite** [comment]
> Is there a significant number of adult porn in VIC category 2 (by your experience)? Or most of them are really sequences or age difficult? If clear adult porn is already found in VIC 2, we already have that ICSE issue happening...

No, I don't think adult porn is common in VIC category 2. Some borderline cases, but not in the same proportion I observed in ICSE.

The false positives I found in ICSE test were mostly nudity/sensual images (not sexual activities) with young adults (mostly women). Some of the files belong to single set with the same model "trying to look younger". So I wouldn't say that most of these cases are not completely unrelated, but strictly speaking, they are not CSAM.

**Comment by lfcnassif** [comment]
Out of curiosity, was victimIdentified flag set for those images? That would be a good indication about adult x child images, I think.

**Comment by tc-wleite** [comment]
> Out of curiosity, was victimIdentified flag set for those images? That would be a good indication about adult x child images, I think.

Good point! I didn't check, but will do that tomorrow.

**Comment by tc-wleite** [comment]
Checking `victimIdentified` flag in the cases I processed, it reduced a bit the proportion of false positives, but less than I expected.
To get a more meaningful comparison, I ran a comparison against ProjectVIC:
 
|Set|Files|Found in VIC|VIC Cat 0|
|:-:|:-:|:-:|:-:|
|Full ICSE|3,113,990|664,184 (21.3%)|902 (0.029%)|
|victimIdentified|1,493,794|407,474 (27.3%)|234 (0.016%)|
|offenderIdentified|1,450,243|410,559 (28.3%)|229 (0.016%)|
|victimIdentified AND offenderIdentified|1,363,839|398,056 (29.2%)|224 (0.016%)|
|victimIdentified OR offenderIdentified|1,580,198|420,017 (26.5%)|239 (0.015%)|

I also processed a larger case, and although I still found a few false positive, the proportion seemed much lower.
So maybe my previous impression was biased.

@lfcnassif, after these tests, I think using `status: pedo` for the whole set is fine.

**Comment by lfcnassif** [comment]
Great, thank you @tc-wleite! Users could also filter by numSubmissions and other flags if they want to, I think this is fine.

**Comment by lfcnassif** [comment]
Closed by #1169
